source/particle/property/composition_reaction.cc

```diff
@@ -0,0 +1,351 @@
+          {
+            const Utilities::NaturalCoordinate<dim> point =
+              this->get_geometry_model().cartesian_to_other_coordinates(particle.get_location(), coordinate_system);
+
+            // Loop over all reaction and compute total change for each compositional field
+            std::vector<double> reactions_for_each_composition (this->n_compositional_fields(), 0.0);
+
+            for (unsigned int i=0; i<reactant_indices.size(); ++i)
+              {
+                std::array<double,2> reactant_and_product;
+                if (reactant_indices[i] == background_index)
+                  reactant_and_product[0] = 0.0;
+                else
+                  reactant_and_product[0] = particle.get_properties()[this->data_position + reactant_indices[i]];
+
+                if (product_indices[i] == background_index)
+                  reactant_and_product[1] = 0.0;
+                else
+                  reactant_and_product[1] = particle.get_properties()[this->data_position + product_indices[i]];
+
+                const double area = reaction_area->value(Utilities::convert_array_to_point<dim>(point.get_coordinates()), i);
+                const double rate = reaction_rate->value(Utilities::convert_array_to_point<2>(reactant_and_product), i);
+
+                // Check if the reaction should occur during the current time step
+                bool reaction_occurs = false;
+                if (this->get_time() > reaction_times[i] &&
+                    (this->get_time() - this->get_timestep()) < reaction_times[i])
+                  reaction_occurs =  true;
+
+                // Subtract the reaction value from the reactant and add it to the product.
+                // For the background, we do not have to do anything.
+                if (reaction_occurs && reactant_indices[i] != background_index)
+                  reactions_for_each_composition[reactant_indices[i]] -= area * rate;
+                if (reaction_occurs && product_indices[i] != background_index)
+                  reactions_for_each_composition[product_indices[i]] += area * rate;
+              }
+
+            // Loop over all compositional fields to apply reactions
+            for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
+              particle.get_properties()[this->data_position + c] += reactions_for_each_composition[c];
+          }
+      }
+
+
+
+      template <int dim>
+      InitializationModeForLateParticles
+      CompositionReaction<dim>::late_initialization_mode () const
+      {
+        return interpolate_respect_boundary;
+      }
+
+
+
+      template <int dim>
+      UpdateTimeFlags
+      CompositionReaction<dim>::need_update() const
+      {
+        return update_time_step;
+      }
+
+
+
+      template <int dim>
+      std::vector<std::pair<std::string, unsigned int>>
+      CompositionReaction<dim>::get_property_information() const
+      {
+        AssertThrow(this->n_compositional_fields() > 0,
```

Comment:
I would move this assert into the `parse_parameters` function. Even if you dont use `n_compositional_fields` there, if this plugin only works with compositions, then we should crash as soon as possible.